lib/src/test/java/io/ably/lib/http/HttpTest.java

```diff
@@ -0,0 +1,447 @@
+
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+    private static RouterNanoHTTPD server;
+
+
+    @BeforeClass
+    public static void setUp() throws IOException {
+        server = new RouterNanoHTTPD(27331);
+        server.addRoute("/status/:code", StatusHandler.class);
+        server.start(NanoHTTPD.SOCKET_READ_TIMEOUT, true);
+
+        while (!server.wasStarted()) {
+            try {
+                Thread.sleep(100);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+    @AfterClass
+    public static void tearDown() {
+        server.stop();
+    }
+
+
+    /*******************************************
+     * Spec: RSC15
+     *******************************************/
+
+
+    /**
+     * <p>
+     * Validates http is not using any fallback host when we receive valid response from http's host
+     * </p>
+     * <p>
+     * Spec: RSC15a
+     * </p>
+     *
+     * @throws Exception
+     */
+    @Test
+    public void http_execute_nofallback() throws Exception {
+        Http http = Mockito.spy(new Http(new ClientOptions(), null));
+
+        String hostExpected = "non.fallback.host";
+        String responseExpected = "Meow";
+        ArgumentCaptor<URL> argument = ArgumentCaptor.forClass(URL.class);
+        http.setHost(hostExpected);
+
+        /* Partially mock http */
+        doReturn(responseExpected)
+                .when(http).httpExecute(argument.capture(), anyString(), aryEq(new Param[0]),
+                any(Http.RequestBody.class), anyBoolean(), any(Http.ResponseHandler.class));
+
+        String responseActual = (String) http.ablyHttpExecute("", "", new Param[0], new Param[0],
+                Mockito.mock(Http.RequestBody.class), Mockito.mock(Http.ResponseHandler.class));
+
+
+        /* Verify
+         *   - http call executed once,
+         *   - with given host,
+         *   - and delivered expected response */
+        verify(http, times(1)).httpExecute(argument.capture(), anyString(), aryEq(new Param[0]),
+                any(Http.RequestBody.class), anyBoolean(), any(Http.ResponseHandler.class));
+        assertThat(argument.getValue().getHost(), is(equalTo(hostExpected)));
+        assertThat(responseActual, is(equalTo(responseExpected)));
+    }
+
+    /**
+     * <p>
+     * Validates http is using a fallback host when HostFailedException thrown
+     * </p>
+     * <p>
+     * Spec: RSC15a
+     * </p>
+     *
+     * @throws Exception
+     */
+    @Test
+    public void http_execute_singlefallback() throws Exception {
+        Http http = Mockito.spy(new Http(new ClientOptions(), null));
+
+        String hostExpectedPattern = PATTERN_HOST_FALLBACK;
+        String responseExpected = "Meow";
+        ArgumentCaptor<URL> argument = ArgumentCaptor.forClass(URL.class);
+        http.setHost("non.fallback.host");
```

Comment:
It turns out it does not pass currently, and I need to update the expected behavior.